community/detectors/papercut_ng_mf_CVE_2023_27350/src/main/java/com/google/tsunami/plugins/papercut/PapercutNGMFVulnDetector.java

```diff
@@ -0,0 +1,142 @@
+        bootstrapModule = PapercutNGMFVulnDetectorBootstrapModule.class)
+// Optionally, each VulnDetector can be annotated by service filtering annotations. For example, if
+// the VulnDetector should only be executed when the scan target is running Jenkins, then add the
+// following @ForSoftware annotation.
+// @ForSoftware(name = "Jenkins")
+public final class PapercutNGMFVulnDetector implements VulnDetector {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+
+  private final Clock utcClock;
+  private final HttpClient httpClient;
+
+  @Inject
+  PapercutNGMFVulnDetector(@UtcClock Clock utcClock, HttpClient httpClient) {
+    this.utcClock = checkNotNull(utcClock);
+    this.httpClient = checkNotNull(httpClient).modify().setFollowRedirects(false).build();
+  }
+
+  // This is the main entry point of your VulnDetector. Both parameters will be populated by the
+  // scanner. targetInfo contains the general information about the scan target. matchedServices
+  // parameter contains all the network services that matches the service filtering annotations
+  // mentioned earlier. If no filtering annotations added, then matchedServices parameter contains
+  // all exposed network services on the scan target.
+  @Override
+  public DetectionReportList detect(
+          TargetInfo targetInfo, ImmutableList<NetworkService> matchedServices) {
+    logger.atInfo().log("CVE-2023-27350 (PaperCut NG/MF) starts detecting.");
+
+    return DetectionReportList.newBuilder()
+            .addAllDetectionReports(
+                    matchedServices.stream()
+                            .filter(NetworkServiceUtils::isWebService)
+                            .filter(this::isServiceVulnerable)
+                            .map(networkService -> buildDetectionReport(targetInfo, networkService))
+                            .collect(toImmutableList()))
+            .build();
+  }
+
+  private boolean isServiceVulnerable(NetworkService networkService) {
+    String rootUri = NetworkServiceUtils.buildWebApplicationRootUrl(networkService);
+    String targetUri = rootUri + "/app?service=page/SetupCompleted";
+    boolean isVulnerable = false;
+
+    HttpHeaders headers = HttpHeaders.builder().addHeader("Origin", rootUri).build();
+
+    HttpRequest req = HttpRequest.post(targetUri).setHeaders(headers).build();
+
+    try {
+      HttpResponse res = httpClient.send(req, networkService);
+      String content = res.bodyString().orElse(null);
+
+      Matcher matches;
+      if (content != null) {
+        matches = Pattern.compile("Configuration Wizard : Setup Complete").matcher(content);
+
+        // if a response code 302 (HttpStatus.FOUND) and/or the title isn't match, then it isn't a
+        // vulnerable version
+        if (res.status() == HttpStatus.OK && matches.find()) {
+          isVulnerable = true;
+        }
+      }
+
+      return isVulnerable;
+    } catch (IOException e) {
```

Comment:
Add `logger.atWarning().withCause(e).log();` here before return for debugging purpose.

Reply:
Added!